sys/kern/pty.c

```diff
@@ -0,0 +1,273 @@
+#include <sys/filedesc.h>
+#include <sys/mutex.h>
+#include <sys/ringbuf.h>
+#include <sys/ttycom.h>
+#include <stdatomic.h>
+#include <stdio.h>
+#include <sys/condvar.h>
+#include <sys/mimiker.h>
+#include <sys/file.h>
+#include <sys/proc.h>
+#include <sys/fcntl.h>
+#include <sys/errno.h>
+#include <sys/stat.h>
+#include <sys/devfs.h>
+#include <sys/linker_set.h>
+#include <sys/tty.h>
+
+#define MAX_PTYS 16
+
+/* Must be greater than or equal to the length of the path to any slave device.
+ * Right now slave devices are at /dev/pts/<number>, so this should suffice. */
+#define PTY_PATH_MAX_LEN 15
+
+static devfs_node_t *pts_dir;
+
+typedef struct {
+  atomic_int pt_number; /* PTY number, if allocated. -1 means free. */
+  condvar_t pt_incv;    /* CV for readers */
+  condvar_t pt_outcv;   /* CV for writers */
+} pty_t;
+
+static pty_t pty_array[MAX_PTYS];
+
+static pty_t *pty_alloc(void) {
+  for (int i = 0; i < MAX_PTYS; i++) {
+    pty_t *pty = &pty_array[i];
+    if (atomic_compare_exchange_strong(&pty->pt_number, &(int){-1}, i + 1))
+      /* Got a free PTY. */
+      return pty;
+  }
+  return NULL;
+}
+
+static void pty_free(pty_t *pty) {
+  pty->pt_number = -1;
+}
```

Comment:
I would add an assert for `pty->pt_number `.

Reply:
Done.